### PR TITLE
feat: ✨ add sd-dialog

### LINKED ISSUE #229: feat: ✨ add sd-dialog
## Description
A dialog appears over other content. It requires an interaction from the user before they can return to whatever is underneath. 

## Documentation
https://www.figma.com/file/WwLYG6UHIUUGGKa59tjGhJ/sd-dialog-V.1.0.0?type=design&node-id=11404-3459&mode=design&t=cmZ7ws0b1ugnBTsn-4
## Hand-off
https://www.figma.com/file/WwLYG6UHIUUGGKa59tjGhJ/sd-dialog-V.1.0.0?type=design&node-id=15256-2501&mode=design&t=cmZ7ws0b1ugnBTsn-4

#### Props
| Status | Name | Type | Default | Description |
| ------------- | ------------- | ------------- | ------------- | ------------- |
| <ul><li>[x] </li></ul>  | `no-close-button` | `boolean` | `false` | removes to close button from the top right => close option (e.g. cancle button) needs to be added into the slot |
| <ul><li>[x] </li></ul>  | `headline` | `string` | '' | to be used for simple text content - rename from shoelace `label` and adopt description |
| <ul><li>[x] </li></ul>  | `noHeader` | N/A | N/A | Remove from component |
| <ul><li>[x] </li></ul>  | `open` | `boolean` | `false` | opens up the dialog |

 - [x] The dialog always has a dark background

#### CSS-Properties
  - [x] `--width` - offers a possibility to set a "responsive" width of the dialog window
  - [x] remove all other properties
#### Parts (besides the shoelace parts)
- [x] `background` - Offers the possibility to change the dark background (`overlay` part already exists)
#### Slots
  - [x] `headline` - used for complex content - rename from shoelace `label` and adopt description | NOTE: its a nested component in Figma
  - [x] `default` - the components content area
  - [x] `footer` - A footer section to add additional elements eg. buttons
  - [x] remove all other slots
#### Stories (besides the `default` story)
  - [x] no-close-button
  - [ ] headline - shows complex headline content => DEV-only
  - [x] scrolling - shows longer content (default slot) and the scrollbar inside the dialog
#### Samples
  - [ ] `default footer` - sample that shows the button in the footer
  - [x] `extended footer` - sample that shows button group with secondary + primary button to show that people should not used the same button variant
  - [x] `small headline` - sample that shows smaller headline sizes (using `<span>` not H2, H3, etc.)
  - [x] `no close button` - sample showing "no-close-button" - closing the dialog using a button in the footer
  - [x] `prevent-closing` - sample showing how to prevent a dialog being closed by clicking on the background

## DoR
- [x] Item has business value
- [x] Item has been estimated by the team
- [x] Item is clear and well-defined
- [x] Item dependencies have been identified

## DoD
- [x] Documentation has been created/updated (if applicable)
- [ ] Migration Guide has been created/updated (if applicable)
- [ ] Relevant E2E tests (Features, A11y, Bug fixes) are created/updated
- [ ] Relevant stories (Features, A11y) are created/updated
- [ ] Implementation works successfully on `feature` branch

## Annotations
Height is flexible according to dialog content. 
Width is to be set via `--width` CSS-property

https://shoelace.style/components/dialog

Compare with current shoelace version on GitHub => our shoelace component is outdated => make sure we update it if needed.

### PR DESCRIPTION
**Pull request by Vahid1919**
## Description:
Added sd-dialog component. Closes #229.

## Definition of Reviewable:
- [x] Documentation is created/updated
- [x] Migration Guide is created/updated
- [x] E2E tests (features, a11y, bug fixes) are created/updated
- [x] Stories (features, a11y) are created/updated
- [x] relevant tickets are linked
- [x] PR is assigned to project board

### CODE REVIEW

File: packages/components/src/components/dialog/dialog.ts [hunk 1]
```diff
@@ -1,33 +1,33 @@
...
- * @dependency sd-icon-button
+ * @dependency sd-button
+ * @dependency sd-icon
  *
  * @slot - The dialog's main content.
- * @slot label - The dialog's label. Alternatively, you can use the `label` attribute.
- * @slot header-actions - Optional actions to add to the header. Works best with `<sd-icon-button>`.
- * @slot footer - The dialog's footer, usually one or more buttons representing various options.
+ * @slot headline - The dialog's headline. Alternatively, you can use the `headline` attribute.
+ * @slot footer - The dialog's footer, usually one or more buttons representing various options. * @slot close-button - The dialog's close button. Works best with `<sd-button>` and `<sd-icon>`.
```
Comment (yoezlem) [inline packages/components/src/components/dialog/dialog.ts:30]:
```suggestion
 * @slot footer - The dialog's footer, usually one or more buttons representing various options. 
 * @slot close-button - The dialog's close button. Works best with `<sd-button>` and `<sd-icon>`.
```

**Comment by yoezlem** [comment]
Somehow the slots also not working?
image

**Comment by yoezlem** [comment]
@Vahid1919 seems like all controls aren't working. Its not only for slots and the boolean for open and close the dialog. Please have a look on that! 👀

**Review by azraefendic — COMMENTED** [review]
Noticing some issues with the controls, let's look into that when you're back. Also what about the stories for Parts, Slots and Mouseless?

**Comment by EllaELSA** [comment]
some minor design issues:
- Version "Headline" for long headlines is the line height too big
- Distance between Headline and Text is bigger than in figma

**Review by EllaELSA — APPROVED** [review]
looks fine for me now

**Review by mariohamann — APPROVED** [review]

**Comment by mariohamann** [comment]
As I expect that there will quite some changes here, I'll move this to draft state to reduce Chromatic load. @Vahid1919 feel free to open it again!

**Comment by Vahid1919** [comment]
> @Vahid1919 seems like all controls aren't working. Its not only for slots and the boolean for open and close the dialog. Please have a look on that! 👀

There seems to be an open issue with Storybook where controls do not work on the Docs page when using iframes (inline=false). (https://github.com/storybookjs/storybook/issues/11908)

Their suggested workaround is to make the stories inline, which is not an option for `sd-dialog` as we won't be able to scroll when a dialog is open making the page unusable.

One option we could consider is replicating [Shoelace's approach](https://shoelace.style/components/dialog) with an "Open Dialog" button on every story. But the whole page will still be unscrollable when 1 dialog is open and some attributes (eg: headline) won't be usable as well. 

@solid-design-system/development None of the solutions I came up with are ideal. Do you have any suggestions in how we can approach this?

**Comment by karlbaumhauer** [comment]
> > @Vahid1919 seems like all controls aren't working. Its not only for slots and the boolean for open and close the dialog. Please have a look on that! 👀
> 
> There seems to be an open issue with Storybook where controls do not work on the Docs page when using iframes (inline=false). ([storybookjs/storybook#11908](https://github.com/storybookjs/storybook/issues/11908))
> 
> Their suggested workaround is to make the stories inline, which is not an option for `sd-dialog` as we won't be able to scroll when a dialog is open, making the page unusable.
> 
> One option we could consider is replicating [Shoelace's approach](https://shoelace.style/components/dialog) with an "Open Dialog" button on every story. But the whole page will still be unscrollable when 1 dialog is open and some attributes (eg: headline) won't be usable as well.
> 
> @solid-design-system/development None of the solutions I came up with are ideal. Do you have any suggestions in how we can approach this?

I would go with an approach, where we might be able to let storybook fix it... Means lets try to find a solution with some buggy behavior which could be solved as soon as storybook updates a bugfix.

**Comment by coraliefeil** [comment]
Approved. Good job on that one @Vahid1919 ;)

**Comment by mariohamann** [comment]
> > @Vahid1919 seems like all controls aren't working. Its not only for slots and the boolean for open and close the dialog. Please have a look on that! 👀
> 
> There seems to be an open issue with Storybook where controls do not work on the Docs page when using iframes (inline=false). ([storybookjs/storybook#11908](https://github.com/storybookjs/storybook/issues/11908))
> 
> Their suggested workaround is to make the stories inline, which is not an option for `sd-dialog` as we won't be able to scroll when a dialog is open, making the page unusable.
> 
> One option we could consider is replicating [Shoelace's approach](https://shoelace.style/components/dialog) with an "Open Dialog" button on every story. But the whole page will still be unscrollable when 1 dialog is open and some attributes (eg: headline) won't be usable as well.
> 
> @solid-design-system/development None of the solutions I came up with are ideal. Do you have any suggestions in how we can approach this?

We could think about providing one single fully interactive story and doing do the rest in style of Shoelace. What do you think?

**Comment by Vahid1919** [comment]
> We could think about providing one single fully interactive story and doing do the rest in style of Shoelace. What do you think?

The Shoelace style might not fully work since we won't be able to make `sd-dialog` fill the entire screen. From what I understand it will be limited to the Canvas. Which is not ideal since it will still block scrolling etc...

**Comment by Vahid1919** [comment]
@coraliefeil I made a small change for the scrolling story that needs approval :)

**Review by mariohamann — COMMENTED** [review]
In general approval: Only some small questions + the question via chat, if a story makes sense to show how to prevent closign etc.

**Review by mariohamann — APPROVED** [review]

File: packages/components/src/components/dialog/dialog.ts [hunk 2]
```diff
@@ -260,55 +252,94 @@ export default class SdDialog extends SolidElement {
...
     `;
   }
+
+  static styles = [
+    componentStyles,
+    unsafeCSS(HeadlineStyles),
+    SolidElement.styles,
+    css`
+      :host {
+        --width: 662px;
```
[the remark this answers is not in the document]
Comment (Vahid1919) [reply packages/components/src/components/dialog/dialog.ts:321]: Yes. I took the exact values from there.

[the remark this answers is not in the document]
Comment (Vahid1919) [reply packages/components/src/components/dialog/dialog.ts:321]: Pulled the values from here:

image

**Comment by coraliefeil** [comment]
Approved.

**Review by azraefendic — APPROVED** [review]

**Review by yoezlem — APPROVED** [review]

**Comment by coraliefeil** [comment]
Approved. THX

**Pull request merged**